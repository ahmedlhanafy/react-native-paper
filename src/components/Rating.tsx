import * as React from 'react';
import { withTheme } from '../core/theming';
import { Theme } from '../types';
import {
  GestureResponderEvent,
  StyleSheet,
  TouchableOpacity,
  View,
} from 'react-native';
import Icon from './Icon';

export type Props = {
  /**
   * Max possible rating value.
   */
  max: number;
  /**
   * Rating value.
   */
  value: number;
  /**
   * Whether the component editable or not.
   */
  editable?: boolean;
  /**
   * Fill colour.
   */
  color?: string;
  /**
   * Size of icons.
   */
  size?: number;
  /**
   * Whether to animate icons on touch.
   */
  animate?: boolean;
  /**
   * Whether half rating input is allowed.
   */
  allowHalfInput?: boolean;
  /**
   * User defined styles for wrapping icons.
   */
  customStyles?: object;
  /**
   * Callback to access changed rating value in editable version.
   */
  onRatingChanged?: ((changedValue: number) => void) | (() => void);
  /**
   * @optional
   */
  theme: Theme;
};

export type State = {
  /**
   * Rating value.
   */
  value: number;
};

/**
 * Rating component for visualizing scores/feedback.
 * It has both a read only and an editable version.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img src="screenshots/rating-default.png" />
 *     <figcaption>Default</figcaption>
 *   </figure>
 *   <figure>
 *     <img src="screenshots/rating-custom-styles.png" />
 *     <figcaption>With Customizations</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 import * as React from 'react';
 import { View } from 'react-native';
 import { Rating } from 'react-native-paper';

 export default class RatingExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      editableValue: 1,
    };
  }

  onRatingChanged(changedValue) {
    this.setState({ editableValue: changedValue });
  }

  render() {
    const { editableValue } = this.state;
    return (
      <View>
        <Rating value={3} max={5} color={'red'} size={24} />
        <Rating value={2.5} max={5} color={'#0c6cc1'} />
        <Rating
          value={editableValue}
          max={10}
          editable={true}
          onRatingChanged={this.onRatingChanged.bind(this)}
        />
        <Rating
          value={1.5}
          max={5}
          color={'magenta'}
          customStyles={{flex: 1, alignItems: 'center'}}
        />
      </View>
    );
  }
}
 * ```
 */
class Rating extends React.Component<Props, State> {
  static defaultProps: Partial<Props> = {
    editable: false,
    color: 'gold',
    size: 36,
    animate: true,
    allowHalfInput: true,
    customStyles: {},
    onRatingChanged: () => {},
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      value: props.max / 2,
    };
  }

  updateRating = (e: GestureResponderEvent, i: number) => {
    const { editable, onRatingChanged } = this.props;
    if (!editable) return;
    const newRating = this.calculateRating(e, i);
    this.setState({ value: newRating });
    // @ts-ignore
    onRatingChanged(newRating);
  };

  calculateRating = (event: GestureResponderEvent, i: number) => {
    const { size, allowHalfInput } = this.props;
    if (!allowHalfInput) return i;
    return event.nativeEvent.locationX < Number(size) / 2 ? i - 0.5 : i;
  };

  getIconSource = (value: number, i: number) => {
    return i > value
      ? i - value <= 0.5
        ? 'star-half'
        : 'star-outline'
      : 'star';
  };

  render() {
    const { max, color, animate, customStyles } = this.props;
    const size = Number(this.props.size);
    const opacity = animate ? 0.5 : 1;
    const { value } = this.state;
    let source;
    let outputs = [];

    for (let i = 1; i <= max; ++i) {
      source = this.getIconSource(value, i);
      outputs.push(
        <TouchableOpacity
          activeOpacity={opacity}
          key={i}
          onPress={e => this.updateRating(e, i)}
          style={customStyles}
        >
          <Icon source={source} color={color} size={size} />
        </TouchableOpacity>
      );
    }
    return <View style={styles.container}>{outputs}</View>;
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
  },
});

export default withTheme(Rating);
