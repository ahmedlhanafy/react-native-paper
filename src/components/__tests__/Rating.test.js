import * as React from 'react';
import renderer from 'react-test-renderer';
import Rating from '../Rating.tsx';
import { TouchableOpacity } from 'react-native';

it('renders rating component by default', () => {
  const tree = renderer.create(<Rating max={5} value={2.5} />).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should render children equal to max value', () => {
  const tree = renderer.create(<Rating max={5} value={2.5} />).toJSON();

  expect(tree.children).toHaveLength(5);
});

it('should call onRatingChanged whenever user changes rating', () => {
  const ratingChanged = jest.fn();
  const event = {
    nativeEvent: {
      locationX: 20,
    },
  };
  const tree = renderer.create(
    <Rating
      max={5}
      value={2.5}
      editable={true}
      onRatingChanged={ratingChanged}
    />
  ).root;

  tree.findAllByType(TouchableOpacity)[2].props.onPress(event);

  expect(ratingChanged).toHaveBeenCalledWith(3);
});

it('should support half star rating', () => {
  const ratingChanged = jest.fn();
  const event = {
    nativeEvent: {
      locationX: 20,
    },
  };
  const tree = renderer.create(
    <Rating
      max={5}
      value={2.5}
      editable={true}
      size={50}
      onRatingChanged={ratingChanged}
    />
  ).root;

  tree.findAllByType(TouchableOpacity)[2].props.onPress(event);

  expect(ratingChanged).toHaveBeenCalledWith(2.5);
});

it('should take full rating when user clicks icon in full', () => {
  const ratingChanged = jest.fn();
  const event = {
    nativeEvent: {
      locationX: 40,
    },
  };
  const tree = renderer.create(
    <Rating
      max={5}
      value={2.5}
      editable={true}
      size={50}
      onRatingChanged={ratingChanged}
    />
  ).root;

  tree.findAllByType(TouchableOpacity)[2].props.onPress(event);

  expect(ratingChanged).toHaveBeenCalledWith(3);
});

it('should not allow user to select rating when editable is false', () => {
  const ratingChanged = jest.fn();
  const event = {
    nativeEvent: {
      locationX: 40,
    },
  };
  const tree = renderer.create(
    <Rating
      max={5}
      value={2.5}
      editable={false}
      size={50}
      animate={false}
      onRatingChanged={ratingChanged}
    />
  ).root;

  tree.findAllByType(TouchableOpacity)[2].props.onPress(event);

  expect(ratingChanged).not.toHaveBeenCalled();
});

it('should not support half star when allowHalf passed false', () => {
  const ratingChanged = jest.fn();
  const event = {
    nativeEvent: {
      locationX: 10,
    },
  };
  const tree = renderer.create(
    <Rating
      max={5}
      value={3}
      editable={true}
      size={50}
      onRatingChanged={ratingChanged}
      allowHalfInput={false}
    />
  ).root;

  tree.findAllByType(TouchableOpacity)[2].props.onPress(event);

  expect(ratingChanged).toHaveBeenCalledWith(3);
});
