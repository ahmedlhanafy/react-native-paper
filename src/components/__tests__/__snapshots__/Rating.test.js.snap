// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders rating component by default 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 36,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 36,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 36,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 36,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text
      accessibilityElementsHidden={true}
      allowFontScaling={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        Array [
          Object {
            "color": "gold",
            "fontSize": 36,
          },
          Array [
            Object {
              "transform": Array [
                Object {
                  "scaleX": 1,
                },
              ],
            },
            Object {
              "backgroundColor": "transparent",
            },
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
